solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateReaderTest.java

```diff
@@ -609,4 +612,60 @@ public void testWatchRaceCondition() throws Exception {
       ExecutorUtil.awaitTermination(executorService);
     }
   }
+
+  /**
+   * Ensure that `fetchCollectionState` would not throw exception when the state.json is deleted in between the
+   * state.json read and PRS entries read
+   * @throws Exception
+   */
+  public void testDeletePrsCollection() throws Exception {
+    ZkStateWriter writer = fixture.writer;
+    ZkStateReader reader = fixture.reader;
+
+    fixture.zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c1", true);
+
+    ClusterState clusterState = reader.getClusterState();
+    // create new collection
+    DocCollection state =
+        new DocCollection(
+            "c1",
+            new HashMap<>(),
+            Collections.singletonMap(DocCollection.PER_REPLICA_STATE, true),
+            DocRouter.DEFAULT,
+            0,
+            ZkStateReader.CLUSTER_STATE + "/c1/state.json");
+    ZkWriteCommand wc = new ZkWriteCommand("c1", state);
+    writer.enqueueUpdate(clusterState, Collections.singletonList(wc), null);
+    writer.writePendingUpdates();
+
+    TimeOut timeOut = new TimeOut(5000, TimeUnit.MILLISECONDS, TimeSource.NANO_TIME);
+    timeOut.waitFor(
+        "Timeout on waiting for c1 to show up in cluster state",
+        () -> reader.getClusterState().getCollectionOrNull("c1") != null);
+
+    String collectionPath = ZkStateReader.getCollectionPath("c1");
+    ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(new SolrNamedThreadFactory("zk-state-reader-test-scheduler"));
+    executor.schedule(() -> {
+      try {
+        reader.getZkClient().clean(collectionPath); //call this directly, very tricky to control execution order with writer.enqueueUpdate
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } catch (KeeperException e) {
+        throw new RuntimeException(e);
+      }
+    },100, TimeUnit.MILLISECONDS);
+
+    CommonTestInjection.setDelay(1000);
+    timeOut.waitFor("Timeout waiting for collection state to become null", () -> {
```

Comment:
This says `Timeout waiting for collection state to become null`, but it looks like we don't actually check that the returned value is null? (It should be, but might be worth explicitly checking `assertNull()` or whatever, if only for the sake of clarity.

Reply:
the lamba passed in should check that it's null? if it's non-null, i suppose this `waitFor` should throw TimeOutException and fail the test case :)